keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md

```diff
@@ -0,0 +1,398 @@
+algorithm (heapsort) that used by the current queue to get the next pods depends on static
+data.
+
+Other requirements come from current semi-exposed functions/parameters:
+For example, the interval of function flushUnschedulableQLeftover, although configurable,
+is immutable after it gets initialized. A requirement is to adjust the interval dynamically
+based on the pressure of pending workloads (e.g., length of queue or other metrics).
+
+Given that the business needs may vary greatly, it'd be desirable to provide a replaceable
+`SchedulerQueue` interface and plumbing mechanics. So that the developers can
+implement their custom queue implementation, while the upstream maintainers focus
+on keeping the core small and extensible, and thus only maintain the internal queue piece.
+
+### Goals
+
+1. Support custom queue implementation of the scheduler.
+2. Roll out the design gradually. In phase 1, keep current internal queue implementation
+intact.
+
+### Non-Goals
+
+Support custom rules for pod selection in the current internal queue of scheduler.
+
+## Proposal
+
+Provide an interface like current scheduler plugins, users can provide a
+custom queue of scheduler and register it, then the kubernetes
+scheduler will use this custom queue for pod management.
+
+This is an extension of the current scheduler plugins, users can control
+more details of the scheduler with this enhancement.
+
+Pros:  
+Users can get full control of the pod queuing logic, they can pop/re-queue/backoff pods with
+custom logic.
+
+The scheduler plugins design is enhanced, the custom queue and the extension
+points can work together to meet more requests.
+
+Cons:  
+Users need to understand the details of the current queue which has 13
+functions now, it is not easy to implement a custom queue.
+
+### Risks and Mitigations
+
+A poor-implemented queue may not function well, in both functionality and performance.
+But the default scheduler works as before and thus won't be impacted.
+
+## Design Details
+
+The scheduler internal queue, as its name implies, was originally implemented for internal
+usage. Due to that, some structs (like cache, Option) are not designed for external extension,
+and the internal queue is initialized with specified parameters like LessFunc,
+SharedInformerFactory, and Option by [factory.go#create](https://github.com/kubernetes/kubernetes/blob/f1f0183d2bbcde33024b2a05d6f39df32f11e037/pkg/scheduler/factory.go#L172). Although refactoring the internal queuing mechanics to be reusable/extensible is the eventual
+goal, we'd like to approach that gradually, and thus is not the goal of phase 1 of this design.
+In phase 1, we'd like to exercise the idea of "a replaceable scheduler queue" in a "minimal viable product"
+manner.
+
+A practical design for phase 1 is described as below:
+
+1. Make `SchedulingQueue` a public interface. Users can choose to implement their own queue.
```

Comment:
The interface definition should be part of the KEP

Reply:
There is no change for the interface `SchedulingQueue`, just make it from private to publish, this is why its details are not listed. To avoid confusing, provide following message for more details:

1. Make `SchedulingQueue` a public interface which is private now in ([scheduling_queue.go](https://github.com/kubernetes/kubernetes/blob/ee459b8969ed2abfed79a07d4ac9d41f13f18ce6/pkg/scheduler/internal/queue/scheduling_queue.go#L126)). Users can choose to implement their own queue.